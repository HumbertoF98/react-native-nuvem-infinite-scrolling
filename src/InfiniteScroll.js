import React, {useEffect} from 'react';
import {View, FlatList} from 'react-native';
import {_} from 'lodash';

const ItemList = ({ showScroll=true, renderRow, data, onLoadMore = ()=> {} }) => {

  useEffect(() => {
    try {
    onLoadMore();
    } catch(error) {
      console.log(error);
    }
  }, []);

  return (
    <View>
      <FlatList
        data={data}
        keyExtractor={incident => String(incident.id)}
        showsVerticalScrollIndicator={showScroll}
        onEndReached={onLoadMore}
        onEndReachedThreshold={0.2}
        renderItem={renderRow}
      />
    </View>
  );
};

export default ItemList;