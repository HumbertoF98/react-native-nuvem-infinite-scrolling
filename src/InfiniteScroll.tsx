import React from 'react';
import {
  View,
  FlatList,
  FlatListProps,
  ActivityIndicator,
  StyleSheet,
  Text,
} from 'react-native';

export interface InfiniteScrollProps<ItemT = any> extends FlatListProps<ItemT> {
  loading?: Boolean;
  loadingText?: string;
  isFinish?: Boolean;
  renderLoading?: Function;
  onLoadMore?: Function;
}

function InfiniteScroll({
  loading = false,
  isFinish = false,
  loadingText,
  renderLoading,
  onLoadMore,
  ...rest
}: InfiniteScrollProps) {
  return (
    <View style={styles.container}>
      <FlatList
        testID="flatlist-infinite-scroll"
        onEndReached={() => {
          if (!loading && typeof onLoadMore === 'function' && !isFinish) {
            onLoadMore();
          }
        }}
        onEndReachedThreshold={0.2}
        {...rest}
      />

      {loading ? (
        typeof renderLoading === 'function' ? (
          renderLoading()
        ) : (
          <View style={styles.loadingWrapper} testID="loading-wrapper">
            <ActivityIndicator size="large" />

            {loadingText ? (
              <Text style={styles.loadingText} testID="loading-text">
                {loadingText}
              </Text>
            ) : null}
          </View>
        )
      ) : null}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  loadingWrapper: {
    marginVertical: 16,
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadingText: {
    marginTop: 8,
  },
});

export default InfiniteScroll;
