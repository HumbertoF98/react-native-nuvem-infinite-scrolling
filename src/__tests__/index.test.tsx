import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import InfiniteScroll from '../InfiniteScroll';
import { View, NativeScrollEvent, Text } from 'react-native';

describe('# InfiniteScroll', () => {
  test('must render', () => {
    const mock = [
      {
        name: 'John',
      },
    ];

    const renderItem = jest.fn();

    const tree = render(<InfiniteScroll data={mock} renderItem={renderItem} />);
    expect(tree).toMatchSnapshot();
  });

  test('must render the items', async () => {
    const heightItem = 150;

    const mock = [
      {
        name: 'John 1',
      },
      {
        name: 'John 2',
      },
      {
        name: 'John 3',
      },
      {
        name: 'John 4',
      },
      {
        name: 'John 5',
      },
      {
        name: 'John 6',
      },
    ];

    const renderItem = ({ item, index }: any) => (
      <View
        testID={`item-${index + 1}`}
        key={index}
        style={{ height: heightItem }}
      >
        <Text>{item.name}</Text>
      </View>
    );

    const onLoadMore = jest.fn();

    const { getByTestId } = render(
      <InfiniteScroll
        data={mock}
        renderItem={renderItem}
        onLoadMore={onLoadMore}
        keyExtractor={(item) => item.name}
      />
    );

    const item1 = getByTestId('item-1');
    const item2 = getByTestId('item-2');
    const item3 = getByTestId('item-3');
    const item4 = getByTestId('item-4');
    const item5 = getByTestId('item-5');
    const item6 = getByTestId('item-6');

    expect(item1).not.toBeNull();
    expect(item2).not.toBeNull();
    expect(item3).not.toBeNull();
    expect(item4).not.toBeNull();
    expect(item5).not.toBeNull();
    expect(item6).not.toBeNull();
  });

  test('should call the load more function', async () => {
    const heightItem = 150;

    const mock = [
      {
        name: 'John 1',
      },
      {
        name: 'John 2',
      },
      {
        name: 'John 3',
      },
      {
        name: 'John 4',
      },
      {
        name: 'John 5',
      },
      {
        name: 'John 6',
      },
    ];

    const totalHeight = heightItem * mock.length;

    const renderItem = ({ item, index }: any) => (
      <View
        testID={`item-${index + 1}`}
        key={index}
        style={{ height: heightItem }}
      >
        <Text>{item.name}</Text>
      </View>
    );

    const onLoadMore = jest.fn();

    const { getByTestId } = render(
      <InfiniteScroll
        data={mock}
        renderItem={renderItem}
        onLoadMore={onLoadMore}
        keyExtractor={(item) => item.name}
      />
    );

    const flatList = getByTestId('flatlist-infinite-scroll');

    const scrollY = 768;

    const nativeEvent: NativeScrollEvent = {
      contentInset: {
        bottom: 0,
        left: 0,
        right: 0,
        top: 0,
      },
      contentOffset: {
        y: 0,
        x: 0,
      },
      contentSize: {
        height: totalHeight,
        width: 480,
      },
      layoutMeasurement: {
        height: scrollY,
        width: 480,
      },
      zoomScale: 1,
    };

    await fireEvent.scroll(flatList, {
      nativeEvent,
    });

    expect(onLoadMore).toHaveBeenCalled();
  });

  test('should not call the load more function', async () => {
    const heightItem = 150;

    const mock = [
      {
        name: 'John 1',
      },
      {
        name: 'John 2',
      },
      {
        name: 'John 3',
      },
      {
        name: 'John 4',
      },
      {
        name: 'John 5',
      },
      {
        name: 'John 6',
      },
    ];

    const totalHeight = heightItem * mock.length;

    const renderItem = ({ item, index }: any) => (
      <View
        testID={`item-${index + 1}`}
        key={index}
        style={{ height: heightItem }}
      >
        <Text>{item.name}</Text>
      </View>
    );

    const onLoadMore = jest.fn();

    const { getByTestId } = render(
      <InfiniteScroll
        data={mock}
        renderItem={renderItem}
        onLoadMore={onLoadMore}
        keyExtractor={(item) => item.name}
      />
    );

    const flatList = getByTestId('flatlist-infinite-scroll');

    const scrollY = 200;

    const nativeEvent: NativeScrollEvent = {
      contentInset: {
        bottom: 0,
        left: 0,
        right: 0,
        top: 0,
      },
      contentOffset: {
        y: 0,
        x: 0,
      },
      contentSize: {
        height: totalHeight,
        width: 480,
      },
      layoutMeasurement: {
        height: scrollY,
        width: 480,
      },
      zoomScale: 1,
    };

    await fireEvent.scroll(flatList, {
      nativeEvent,
    });

    expect(onLoadMore).not.toHaveBeenCalled();
  });

  test('must present a loading', () => {
    const mock = [
      {
        name: 'John',
      },
    ];

    const renderItem = jest.fn();

    const { getByTestId } = render(
      <InfiniteScroll data={mock} loading={true} renderItem={renderItem} />
    );

    const loadingElement = getByTestId('loading-wrapper');

    expect(loadingElement).toBeDefined();
  });

  test('must display a loading text', () => {
    const renderItem = jest.fn();

    const { getByTestId } = render(
      <InfiniteScroll
        data={[]}
        loadingText="Loading"
        loading={true}
        renderItem={renderItem}
      />
    );

    const loadingTextElement = getByTestId('loading-text');

    expect(loadingTextElement).toBeDefined();
    expect(loadingTextElement.props.children).toBe('Loading');
  });
});
