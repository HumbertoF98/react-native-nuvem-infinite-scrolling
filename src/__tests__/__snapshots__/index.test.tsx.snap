// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# InfiniteScroll must render 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "John",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flatlist-infinite-scroll"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
</View>
`;
